crates/net/network/src/transactions/mod.rs

```diff
@@ -1091,113 +1111,172 @@ where
+                                    // transaction with the assumption that the peer should have
+                                    // known that this transaction is bad. (e.g. consensus
+                                    // rules)
+                                    if err.is_bad_transaction() && !this.network.is_syncing() {
+                                        debug!(target: "net::tx", ?err, "bad pool transaction import");
+                                        this.on_bad_import(err.hash);
+                                        continue
+                                    }
+                                    this.on_good_import(err.hash);
+                                }
+                            }
+                        }
+
+                        some_ready = true;
+                    }
+                },
+                acc_pending_imports
+            );
+
+            metered_poll!(
+                {
+                    // handle and propagate new transactions.
+                    //
+                    // higher priority! stream is drained
+                    //
+                    let mut new_txs = Vec::new();
+                    while let Poll::Ready(Some(hash)) =
+                        this.imported_transactions.poll_next_unpin(cx)
+                    {
+                        new_txs.push(hash);
+                    }
+                    if !new_txs.is_empty() {
+                        this.on_new_transactions(new_txs);
+                    }
+                },
+                acc_imported_txns
+            );
 
             // all channels are fully drained and import futures pending
             if !some_ready {
-                return Poll::Pending
+                break
             }
 
             budget -= 1;
             if budget <= 0 {
                 // Make sure we're woken up again
                 cx.waker().wake_by_ref();
-                return Poll::Pending
+                break
             }
         }
+
+        // update metrics for whole poll function
+        this.metrics.duration_poll_tx_manager.set(start.elapsed());
+        // update poll metrics for nested streams
+        this.metrics.acc_duration_poll_network_events.set(acc_network_events.as_micros() as f64);
+        this.metrics
+            .acc_duration_poll_pending_pool_imports
+            .set(acc_pending_imports.as_micros() as f64);
+        this.metrics.acc_duration_poll_transaction_events.set(acc_tx_events.as_micros() as f64);
+        this.metrics
+            .acc_duration_poll_imported_transactions
+            .set(acc_imported_txns.as_micros() as f64);
+        this.metrics.acc_duration_poll_fetch_events.set(acc_fetch_events.as_micros() as f64);
+        this.metrics.acc_duration_fetch_pending_hashes.set(acc_pending_fetch.as_micros() as f64);
+        this.metrics.acc_duration_poll_commands.set(acc_cmds.as_micros() as f64);
```

Comment:
just noticed: these should all use `as_secs_f64` - makes it easy for grafana to handle granularity, ie ms, us, ns

Reply:
I can't get this to work well in grafana, maybe since I am running the grafana version we ship in docker and you are running a newer one? when unit is then set to seconds fort the axis, I get huge differences in magnitude, look at these two time ranges, -24h until now and below -24h to -12h. that doesn't seem right.

<img width="1412" alt="Screenshot 2024-02-22 at 11 06 38" src="https://github.com/paradigmxyz/reth/assets/58548332/e565628d-17ce-403a-be6c-4befb1d0dfa9">

<img width="1412" alt="Screenshot 2024-02-22 at 11 07 01" src="https://github.com/paradigmxyz/reth/assets/58548332/8026205b-cd99-4090-bb2a-3fdc6d7e1158">

Reply:
yeah I know it occurred to me too. solved. then green light on these metrics, they can be seen as source of truth.